block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala

```diff
@@ -17,34 +18,117 @@ import coop.rchain.blockstorage.util.BlockMessageUtil.{blockNumber, bonds, paren
 import coop.rchain.blockstorage.util.{BlockMessageUtil, Crc32, TopologicalSortUtil}
 import coop.rchain.blockstorage.util.byteOps._
 import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.catscontrib.Capture
 import coop.rchain.crypto.codec.Base16
-import coop.rchain.shared.{Log, LogSource}
+import coop.rchain.shared.{AtomicMonadState, Log, LogSource}
+import monix.execution.atomic.AtomicAny
 
 import scala.ref.WeakReference
 import scala.util.matching.Regex
 import collection.JavaConverters._
 
+private final case class BlockDagFileStorageState[F[_]: Sync](
+    latestMessages: Map[Validator, BlockHash],
+    childMap: Map[BlockHash, Set[BlockHash]],
+    dataLookup: Map[BlockHash, BlockMetadata],
+    topoSort: Vector[Vector[BlockHash]],
+    sortOffset: Long,
+    checkpoints: List[Checkpoint],
+    latestMessagesLogOutputStream: OutputStream,
+    latestMessagesLogSize: Int,
+    latestMessagesCrc: Crc32[F],
+    blockMetadataLogOutputStream: OutputStream,
+    blockMetadataCrc: Crc32[F]
+)
+
 final class BlockDagFileStorage[F[_]: Concurrent: Sync: Log: BlockStore] private (
     lock: Semaphore[F],
-    latestMessagesRef: Ref[F, Map[Validator, BlockHash]],
-    childMapRef: Ref[F, Map[BlockHash, Set[BlockHash]]],
-    dataLookupRef: Ref[F, Map[BlockHash, BlockMetadata]],
-    topoSortRef: Ref[F, Vector[Vector[BlockHash]]],
-    sortOffsetRef: Ref[F, Long],
-    checkpointsRef: Ref[F, List[Checkpoint]],
-    latestMessagesLogOutputStreamRef: Ref[F, OutputStream],
-    latestMessagesLogSizeRef: Ref[F, Int],
-    latestMessagesCrcRef: Ref[F, Crc32[F]],
     latestMessagesDataFilePath: Path,
     latestMessagesCrcFilePath: Path,
     latestMessagesLogMaxSizeFactor: Int,
-    blockMetadataLogOutputStreamRef: Ref[F, OutputStream],
-    blockMetadataCrcRef: Ref[F, Crc32[F]],
     blockMetadataLogPath: Path,
-    blockMetadataCrcPath: Path
+    blockMetadataCrcPath: Path,
+    state: MonadState[F, BlockDagFileStorageState[F]]
```

Comment:
is this code is supposed to be run by multiple threads?

Reply:
re @rabbitonweb's question:

> regarding `Ref` -> `MonadState` - is this code will be run by multiple threads? If so, consider using `Cell` instead.

I'd say let's not unless you need the `modifyF/flatModify` that typeclass exposes additionally. The code compiles, so you don't :) Let's use the weakest typeclass possible, as long as it's possible :)